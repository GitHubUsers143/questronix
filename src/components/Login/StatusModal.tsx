import { View, Text, StyleSheet, Modal, Pressable } from 'react-native';
import React from 'react';
import { COLORS } from '../../constants/colors';

interface StatusModalInterface {
  showing: boolean;
  setShowing: React.Dispatch<React.SetStateAction<boolean>>;
  message: string | undefined;
  description: string | undefined;
  buttonText: string | undefined;
}

const StatusModal: React.FC<StatusModalInterface> = ({
  showing,
  setShowing,
  message = 'Something Went Wrong!',
  description = 'Something went wrong while trying to do something, please try again.',
  buttonText = 'OK',
}) => {
  const closeModal = () => {
    setShowing(false);
  };

  return (
    <Modal
      testID='status-modal'
      animationType='slide'
      transparent={true}
      visible={showing}
      onRequestClose={() => {
        closeModal();
      }}
    >
      <View style={styles.centeredView}>
        <View style={styles.modalView}>
          <Text style={styles.modalTextHeader}>{message.toUpperCase()}</Text>
          <Text>{description}</Text>
          <Pressable
            testID='status-modal-button'
            style={[styles.button, styles.buttonClose]}
            onPress={() => closeModal()}
          >
            <Text style={styles.closeTextStyle}>
              {buttonText.toUpperCase()}
            </Text>
          </Pressable>
        </View>
      </View>
    </Modal>
  );
};

const styles = StyleSheet.create({
  centeredView: {
    flex: 1,
    justifyContent: 'center',
    marginTop: 22,
  },
  modalView: {
    margin: 20,
    backgroundColor: COLORS.WHITE,
    paddingTop: 15,
    paddingBottom: 5,
    paddingLeft: 20,
    paddingRight: 20,
    shadowColor: COLORS.BLACK,
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 4,
    elevation: 5,
  },
  button: {
    borderRadius: 20,
    padding: 10,
    elevation: 2,
  },
  buttonClose: {
    backgroundColor: COLORS.WHITE,
  },
  closeTextStyle: {
    color: COLORS.BLUE,
    fontWeight: 'bold',
    textAlign: 'right',
    marginBottom: 5,
  },
  modalTextHeader: {
    color: COLORS.BLUE,
    fontWeight: 'bold',
    marginBottom: 5,
  },
});

export default StatusModal;
