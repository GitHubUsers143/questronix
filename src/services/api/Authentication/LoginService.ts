import { httpRequest, setToken } from '../../../library/api';

const loginService = async ({
  email,
  password,
}: {
  email: string;
  password: string;
}) => {
  return httpRequest({
    url: 'v1/login-api',
    method: 'POST',
    data: {
      email: email,
      password: password,
    },
  })
    .then(async (response) => {
      await setToken(response.token);
      return Promise.resolve(response.token);
    })
    .catch((error) => {
      return Promise.reject(error || {});
    });
};

const googleSignInService = async ({
  id,
  email,
  name,
  token,
}: {
  id: string;
  email: string;
  name: string | null;
  token: string | null;
}) => {
  return httpRequest({
    url: 'v1/google-login',
    method: 'POST',
    data: {
      id: id,
      email: email,
      name: name,
      token: token,
    },
  })
    .then(async (response) => {
      console.log(response);
      await setToken(response.token);
      return Promise.resolve(response);
    })
    .catch((error) => {
      return Promise.reject(error || {});
    });
};

export default { loginService, googleSignInService };
