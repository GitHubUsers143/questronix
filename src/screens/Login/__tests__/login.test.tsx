import * as React from 'react';
import { fireEvent, render, screen } from '@testing-library/react-native';
import LoginScreen from '../Login';
import { act } from 'react-test-renderer';
import {
  minPassword,
  testEmail,
  testHash,
  testPass,
  validationMessage,
} from '../../../constants/loginValidationMessages';
import * as service from '../../../services/api/Authentication/LoginService';
import StatusModal from '../../../components/Login/StatusModal';
import {
  failButtonText,
  failDescription,
  failMessage,
} from '../../../constants/loginFailMessages';

beforeEach(async () => {
  render(<LoginScreen />);
});

afterEach(async () => {});

describe('user login', () => {
  describe('when user enters incorrect email', () => {
    it('renders invalid email address message', async () => {
      const emailInput = screen.getByTestId('email-input');
      fireEvent.changeText(emailInput, 'test');

      const emailErrorOutput = await screen.findByTestId('email-error');
      expect(emailErrorOutput.props.children).toMatch(
        validationMessage.emailInvalid
      );
    });
  });

  describe('when user empties the email input', () => {
    it('renders email address required message', async () => {
      const emailInput = screen.getByTestId('email-input');
      fireEvent.changeText(emailInput, '');

      const emailErrorOutput = await screen.findByTestId('email-error');
      expect(emailErrorOutput.props.children).toBe(
        validationMessage.emailRequired
      );
    });
  });

  describe('when user empties the password input', () => {
    it('renders password required message', async () => {
      const passwordInput = screen.getByTestId('password-input');
      fireEvent.changeText(passwordInput, '');

      const passwordErrorOutput = await screen.findByTestId('password-error');
      expect(passwordErrorOutput.props.children).toBe(
        validationMessage.passwordRequired
      );
    });
  });

  describe(`when user enters password character less than ${minPassword}`, () => {
    it('renders minimum character password message', async () => {
      const passwordInput = screen.getByTestId('password-input');
      fireEvent.changeText(passwordInput, '12');

      const passwordErrorOutput = await screen.findByTestId('password-error');
      expect(passwordErrorOutput.props.children).toBe(
        validationMessage.passwordMinRequired
      );
    });
  });

  describe('when user enters correct email and password', () => {
    it('renders login success', async () => {
      const emailInput = screen.getByTestId('email-input');
      const passwordInput = screen.getByTestId('password-input');

      await act(() => {
        fireEvent.changeText(emailInput, testEmail);
        fireEvent.changeText(passwordInput, testPass);
      });
      const loginButton = screen.getByTestId('login-button');

      expect(loginButton.props.accessibilityState.disabled).not.toBeTruthy();
    });

    describe('when user clicks the login button', () => {
      it('renders login success', async () => {
        const mockData = { email: testEmail, password: testPass };

        jest.spyOn(service, 'default').mockResolvedValueOnce(testHash);
        const response = await service.default(mockData);

        expect(response).toEqual(testHash);
      });
    });
  });

  describe('when user enters incorrect email and password', () => {
    it('renders login fail', async () => {
      const emailInput = screen.getByTestId('email-input');
      const passwordInput = screen.getByTestId('password-input');
      await act(() => {
        fireEvent.changeText(emailInput, testEmail);
        fireEvent.changeText(passwordInput, testPass);
      });
      const loginButton = screen.getByTestId('login-button');

      expect(loginButton.props.accessibilityState.disabled).not.toBeTruthy();
    });

    describe('when user clicks the login button', () => {
      it('shows the status modal with a message', async () => {
        const { getByTestId } = render(
          <StatusModal
            showing={true}
            setShowing={() => {}}
            message={failMessage}
            description={failDescription}
            buttonText={failButtonText}
          />
        );

        expect(getByTestId('status-modal').props.visible).toBeTruthy();
      });
    });

    describe('when user clicks the close button on the modal', () => {
      it('closes the modal', () => {
        const handleShowFail = jest.fn();
        const { getByTestId } = render(
          <StatusModal
            showing={true}
            setShowing={handleShowFail}
            message={failMessage}
            description={failDescription}
            buttonText={failButtonText}
          />
        );

        fireEvent.press(getByTestId('status-modal-button'));

        expect(handleShowFail).toHaveBeenCalledTimes(1);
      });
    });
  });

  describe('when all test have been completed', () => {
    it('compares current file to snapshot', async () => {
      expect(screen.toJSON()).toMatchSnapshot();
    });
  });
});
