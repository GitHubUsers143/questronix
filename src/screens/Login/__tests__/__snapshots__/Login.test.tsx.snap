// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user login when all test have been completed compares current file to snapshot 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#242938",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
  >
    <Modal
      animationType="slide"
      hardwareAccelerated={false}
      onRequestClose={[Function]}
      testID="status-modal"
      transparent={true}
      visible={false}
    />
    <View
      style={
        {
          "alignContent": "center",
          "flexDirection": "column",
          "justifyContent": "center",
          "margin": 0,
        }
      }
    />
    <View
      style={
        {
          "paddingBottom": 64,
        }
      }
    >
      <Text
        style={
          {
            "color": "#FFFFFF",
            "fontSize": 18.8,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        LOG IN
      </Text>
    </View>
    <View
      style={
        {
          "width": 660,
        }
      }
    >
      <View
        style={
          {
            "paddingBottom": 37.5,
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "#FFFFFF",
              "borderRadius": 14,
              "height": 93.5,
              "paddingHorizontal": 10,
              "width": "100%",
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "alignItems": "center",
                "borderBottomWidth": 0,
                "borderColor": "#86939e",
                "flexDirection": "row",
                "height": 93.5,
                "paddingLeft": 26.5,
                "paddingRight": 26.5,
                "transform": [
                  {
                    "translateX": 0,
                  },
                ],
              }
            }
          >
            <TextInput
              editable={true}
              keyboardType="email-address"
              onBlur={[Function]}
              onChangeText={[Function]}
              placeholder="Email"
              placeholderTextColor="#86939e"
              style={
                {
                  "color": "#242424",
                  "flex": 1,
                  "fontSize": 18.9,
                  "minHeight": 40,
                }
              }
              testID="email-input"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
          <Text
            style={
              {
                "color": "#ff190c",
                "fontSize": 12,
                "margin": 5,
              }
            }
          />
        </View>
      </View>
      <View
        style={
          {
            "paddingBottom": 37.5,
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "#FFFFFF",
              "borderRadius": 14,
              "height": 93.5,
              "paddingHorizontal": 10,
              "width": "100%",
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "alignItems": "center",
                "borderBottomWidth": 0,
                "borderColor": "#86939e",
                "flexDirection": "row",
                "height": 93.5,
                "paddingLeft": 26.5,
                "paddingRight": 26.5,
                "transform": [
                  {
                    "translateX": 0,
                  },
                ],
              }
            }
          >
            <TextInput
              editable={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              placeholder="Password"
              placeholderTextColor="#86939e"
              secureTextEntry={true}
              style={
                {
                  "color": "#242424",
                  "flex": 1,
                  "fontSize": 18.9,
                  "minHeight": 40,
                }
              }
              testID="password-input"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
          <Text
            style={
              {
                "color": "#ff190c",
                "fontSize": 12,
                "margin": 5,
              }
            }
          />
        </View>
      </View>
      <View
        style={
          {
            "paddingTop": 9.5,
          }
        }
      >
        <View
          style={
            [
              {
                "borderRadius": 3,
                "overflow": "hidden",
              },
              {
                "borderRadius": 3,
              },
              [
                {
                  "borderRadius": 14,
                },
                {
                  "width": 660,
                },
              ],
              false,
            ]
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": false,
                "disabled": true,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="login-button"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "hsl(208, 8%, 90%)",
                  "borderColor": "#2089dc",
                  "borderRadius": 3,
                  "borderWidth": 0,
                  "flexDirection": "row",
                  "height": 93.5,
                  "justifyContent": "center",
                  "padding": 8,
                }
              }
            >
              <Text
                style={
                  {
                    "color": "hsl(208, 8%, 63%)",
                    "fontSize": 18.8,
                    "fontWeight": "bold",
                    "paddingVertical": 1,
                    "textAlign": "center",
                  }
                }
              >
                Login
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
